RFC-0004-jdbc-join-pushdown.md

```diff
@@ -0,0 +1,513 @@
+
+* visitJoinNode to identify which Join nodes can be pushed down and handled by the connector. Multiple JoinNodes need to handle here.
+
+* If the connector, identifies a JoinNode that it wants to rewrite, collapsing the JoinNode into a new TableScanNode property by removing the JoinNode.
+
+* Handle join query or  parameters for join query on new TableScanNode property may be a new data structure on JdbcTableHandle
+
+* Handle output projection for multiple table (from collapsed Join) using new TableScanNode(JdbcTableHandle)
+
+* Modify Jdbc split to handle the new TableScanNode property
+
+* Create Querybuilder to create join query from the new JdbcSplit property
+
+* Handle where clause associated with JoinNode
+
+* Handle other predicate like Limit, rowcount, nvl, abs(), etc.
+
+##  1) JoinNode make available on ConnectorPlanOptimizer
+
+Currently we are focusing only for Jdbc join pushdown. The join pushdown should be at connector level.  If we use join queries from multiple connector like iceberg, postgres and db2 then, 
+
+  * No join pushdown is expected for iceberg tables,
+  * If there are multiple tables to join on postgres which satisfies 'join pushdown condition' then that join should pushed down to 
+    connector as a single postgres join query 
+  * and if there are multiple tables on db2 connector which satisfies 'join pushdown condition' then that also need to pushed down to db2 
+    connector as a single join query. 
+
+Now the presto is responsible for return the final result by joining above iceberg, postgres and db2 connector response.
+
+For implementing connector level join pushdown, first we need JoinNode and make available it for accessing on connector level. For that JoinNode should move to spi package (com.facebook.presto.spi.plan) from sql planner (com.facebook.presto.sql.planner.plan). It includes,
+
+* Moving dependent class to spi package.
+```
+         JoinNode.java
+
+         AbstarctJoinNode.java
+
+         SemiJoinNode.java
+```
+* Refactor codes that use above class : Refactor the imports 
+
+* Currently visitJoin() is implemented on InternalPlanVistor. This class should not move to spi, instead add visit join method on PlanVistor 
+  and extend PlanVistor on JoinNode instead of InternalPlanVistor
+
+* Correct all build fails, test fails and method missing by implementing them for new JoinNode.
+
+* Do sample unit testing and ensure Join query coverage.
+
+## 2) On ApplyConnectorOptimization Optimizer, add new ConnectorPlanOptimizer called JdbcJoinPushdown
+
+We  have ApplyConnectorOptimization Optimizer to add connector specific plan optimizer. This Optimizer is created on application initialization and load values while read catalogue properties. Existing general flow for optimizer is as follows. 
+
+![ConnectorPlanOptimizer](https://github.com/Ajas-Mangal/jdbc-join-pushdown/assets/175085180/7ff937c4-f3d8-40c1-acb8-80c540749f90)
+
+
+#### New Changes required as part of this enhancement
```

Comment:
Do you have these in a prototype PR ? It would be easier to understand from it.

Reply:
Yes. This diagram added in the PR